app/models/filter_tagging.rb

```diff
@@ -1,89 +1,21 @@
 # This is essentially a mirror of the taggings table as applied to works (right now)
 # except with all works connected to canonical tags instead of their synonyms for
 # browsing and filtering purposes. Filter = tag, filterable = thing that's been tagged.
+#
+# Note that this doesn't have very many validations or callbacks -- this is
+# because the ONLY class that should be adding/updating FilterTaggings is the
+# FilterUpdater, which handles those itself.
 class FilterTagging < ApplicationRecord
-  self.primary_key = 'id'
+  belongs_to :filter, class_name: 'Tag', inverse_of: :filter_taggings
+  belongs_to :filterable, polymorphic: true, inverse_of: :filter_taggings
 
-  belongs_to :filter, class_name: 'Tag' # , dependent: :destroy # TODO: poke this separately
-  belongs_to :filterable, polymorphic: true
-
-  validates_presence_of :filter, :filterable
-
-  before_destroy :expire_caches
-  after_create :update_pseud_index
-  after_destroy :update_pseud_index
-
-  def update_pseud_index
-    return unless filter.is_a?(Fandom) && filterable.is_a?(Work)
-    IndexQueue.enqueue_ids(Pseud, filterable.pseud_ids, :background)
-  end
-
-  def self.find(*args)
-    raise "id is not guaranteed to be unique. please install composite_primary_keys gem and set the primary key to id,filter_id"
-  end
-  def self.find_by(id: id_arg)
-    raise "id is not guaranteed to be unique. please install composite_primary_keys gem and set the primary key to id,filter_id"
-  end
+  validates_uniqueness_of :filter_id, scope: [:filterable_type, :filterable_id]
```

Comment:
I agree that it would be good to have a DB constraint for `filter_taggings`.

The `meta_taggings` constraint was added in #3585, so that's handled. The `common_taggings` constraint is [sort of enforced already, but not exactly](https://otwarchive.atlassian.net/browse/AO3-5700).

Reply:
Nice, thanks! I was looking at a very outdated output of consistency_fail.